### PR TITLE
Add Service Discovery classes

### PR DESCRIPTION
**Pull request by awegrzyn**
```
o2::quality_control::core::ServiceDiscovery server{"http://consul:8500", "agent123"};
server._register("task1,task2,task3");
```

### CODE REVIEW

File: Framework/src/ServiceDiscovery.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  public:
+    /// Sets up CURL
+    ServiceDiscovery(const std::string& url, const std::string& id);
+
+    /// Stops the health thread and deregisteres from Consul health checks
+    ~ServiceDiscovery();
+
+    /// Registeres service and health check by sending HTTP PUT request to Consul server
+    /// Provides list of published tasks as Consul tags
+    void _register(const std::string& objects, const std::string& healthUrl);
```
Comment (Barthelemy) [inline Framework/src/ServiceDiscovery.h]: why the underscore ?

File: Framework/src/ServiceDiscovery.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    static void deleteCurl(CURL * curl);
+
+    /// CURL instance
+    std::unique_ptr<CURL, decltype(&ServiceDiscovery::deleteCurl)> curlHandle;
+
+    const std::string consulUrl; ///< Consul URL
+    const std::string id; ///< Instance (service) ID
+    std::thread healthThread; ///< Health check thread
+    std::atomic<bool> mThreadRunning; ///< Health check thread running flag
+    CURL* initCurl(); ///< Initilizes CURL
```
Comment (Barthelemy) [inline Framework/src/ServiceDiscovery.h]: typo

**Comment by Barthelemy** [comment]
We have to remember to add a dependency to curl in the alidist recipe. 
Do you need me to test it or do you want me to wait until it is ready for review ?

File: Framework/src/ServiceDiscovery.h (hunk 1 above)
Comment (awegrzyn) [reply Framework/src/ServiceDiscovery.h]: `register` is apparently keyword in C++

**Comment by awegrzyn** [comment]
@Barthelemy It's all yours.

**Comment by knopers8** [comment]
There seems to be also a legit problem with
```
[100%] Linking CXX executable ../bin/serviceDiscovery
/lib/../lib64/crt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
collect2: error: ld returned 1 exit status
gmake[2]: *** [bin/serviceDiscovery] Error 1
gmake[1]: *** [Framework/CMakeFiles/serviceDiscovery.dir/all] Error 2
gmake: *** [all] Error 2
```
You have put serviceDiscovery to the executables list in cmake and as far as I can see, it is just a class.

**Comment by Barthelemy** [comment]
I can use it and I will make a commit to this PR. However, after I register the service I don't see it in consul UI.

**Comment by Barthelemy** [comment]
ServiceDiscovery constructor takes three arguments. What is the last one for ?

**Comment by awegrzyn** [comment]
> ServiceDiscovery constructor takes three arguments. What is the last one for ?

Local TCP endpoint that Consul connects to in order to perform health check. By default it's set to `<hostname>:7777`. I will document that.

File: Framework/include/QualityControl/ServiceDiscovery.h [hunk 3]
```diff
@@ -15,20 +15,27 @@ namespace o2::quality_control::core
...
 class ServiceDiscovery
 {
   public:
-    /// Sets up CURL
+    /// Sets up CURL and health check
+    /// \param url 		Consul URL
+    /// \param id 		Unique instance ID
+    /// \param healthEndpoint	Local endpoint that is then used for health checks
+    ///				(default value it set to  <hostname>:7777)
     ServiceDiscovery(const std::string& url, const std::string& id, const std::string& healthEndpoint);
```
Comment (Barthelemy) [inline Framework/include/QualityControl/ServiceDiscovery.h]: you say that it has a default value, but does it ? do you mean that I can pass an empty string ?

File: Framework/include/QualityControl/ServiceDiscovery.h [hunk 4]
```diff
@@ -15,20 +15,27 @@ namespace o2::quality_control::core
...
 
     /// Stops the health thread and deregisteres from Consul health checks
     ~ServiceDiscovery();
 
-    /// Registeres service and health check by sending HTTP PUT request to Consul server
-    /// Provides list of published tasks as Consul tags
+    /// Registeres list of online objects by sending HTTP PUT request to Consul server
+    /// \param objects 		List of comma separated tasks
```
Comment (Barthelemy) [inline Framework/include/QualityControl/ServiceDiscovery.h]: "separated objects" ;)

File: Framework/include/QualityControl/ServiceDiscovery.h (hunk 3 above)
Comment (Barthelemy) [reply Framework/include/QualityControl/ServiceDiscovery.h]: just to clarify, if I don't pass the third argument and try to make_unique a ServiceDiscovery there is an error.

**Comment by Barthelemy** [comment]
I am able to see the task running and publishing objects. I have to modify a number of things on our side to publish the list of objects at start of run, but nothing difficult. 
There is a glitch though. The DPL driver is stuck when I close it. The QC Task becomes defunct. No idea why. It is triggered by add the ServiceDiscovery. 
If you want to try it, I can commit the code to use ServiceDiscovery (we can remove the commit later).

**Comment by Barthelemy** [comment]
I have put the code

**Comment by awegrzyn** [comment]
@Barthelemy What's missing to have it merged?

File: Framework/include/QualityControl/ServiceDiscovery.h (hunk 3 above)
Comment (Barthelemy) [reply Framework/include/QualityControl/ServiceDiscovery.h]: See here

**Comment by Barthelemy** [comment]
It is still marked as a draft so I did not merge it :)
Also there is 1 comment that has not been addressed. 
Finally, I will remove the code that use it and make a separate PR.

**Comment by awegrzyn** [comment]
Ok, It's done now

**Comment by Barthelemy** [comment]
I wait for the builds and then merge.

**Pull request merged**